core/src/main/scala/kafka/server/KafkaServer.scala

```diff
@@ -668,7 +684,7 @@ class KafkaServer(
 
           // 1. Find the controller and establish a connection to it.
           // If the controller id or the broker registration are missing, we sleep and retry (if there are remaining retries)
-          metadataCache.getControllerId match {
+          metadataCache.getControllerId.filter(_.isInstanceOf[ZkCachedControllerId]).map(_.id) match {
```

Comment:
If we leave this as `metadataCache.getControllerId match`, we can just add the type to the cases below. E.g., 
```
case Some(controllerId: ZkCachedControllerId) => ...
```